packages/core/src/agent/didcomm/versions/v2/DIDCommV2BaseMessage.ts

```diff
@@ -0,0 +1,164 @@
+
+export type DIDComV2BaseMessageConstructor = Constructor<DIDCommV1BaseMessage>
+
+export type DIDCommV2MessageParams = {
+  type?: string
+  id?: string
+  from?: string
+  to?: string | string[]
+  thid?: string
+  pthid?: string
+  created_time?: number
+  expires_time?: number
+  from_prior?: string
+  attachments?: Array<Attachment>
+  body?: unknown
+}
+
+export class DIDCommV2BaseMessage {
+  @Expose({ name: 'id' })
+  @Matches(MessageIdRegExp)
+  public id!: string
+
+  @Expose({ name: 'type' })
+  @Matches(MessageTypeRegExp)
+  public readonly type!: string
+  public static readonly type: ParsedMessageType
+
+  @Expose({ name: 'typ' })
+  public readonly typ = DIDCommV2BaseMessage.typ
+  public static readonly typ = 'application/didcomm-plain+json'
+
+  @Expose({ name: 'from' })
+  @IsString()
+  @IsOptional()
+  public from?: string
+
+  @Expose({ name: 'to' })
+  @IsArray()
+  @IsOptional()
+  public to?: Array<string>
+
+  @IsNumber()
+  @IsOptional()
+  public created_time?: number
+
+  @IsNumber()
+  @IsOptional()
+  public expires_time?: number
+
+  @IsString()
+  @IsOptional()
+  public thid?: string
+
+  @IsString()
+  @IsOptional()
+  public pthid?: string
+
+  @IsString()
+  @IsOptional()
+  public from_prior?: string
+
+  public body!: unknown
+
+  @IsOptional()
+  public attachments?: Array<Attachment>
+
+  public constructor(options?: DIDCommV2MessageParams) {
+    if (options) {
+      this.id = options.id || this.generateId()
+      this.type = options.type || this.type
+      this.from = options.from
+      this.to = typeof options.to === 'string' ? [options.to] : options.to
+      this.thid = options.thid
+      this.pthid = options.pthid
+      this.created_time = options.created_time
+      this.expires_time = options.expires_time
+      this.from_prior = options.from_prior
+      this.attachments = options.attachments
+      this.body = options.body || {}
+    }
+  }
+
+  public generateId() {
+    return uuid()
+  }
+
+  public static createJSONAttachment<T>(id: string, message: T): Attachment {
+    return {
+      id: id,
+      media_type: ATTACHMENT_MEDIA_TYPE,
+      data: {
+        json: message,
+      },
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public static createBase64Attachment<T>(id: string, message: T): Attachment {
+    return {
+      id: id,
+      media_type: ATTACHMENT_MEDIA_TYPE,
+      data: {
+        base64: JsonEncoder.toBase64(message),
+      },
+    }
+  }
+
+  public static unpackAttachmentAsJson(attachment: Attachment) {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const data = attachment.data as any //FIXME: didcomm package doesn't provide convenient way to process attachment
+    if (typeof data.base64 === 'string') {
+      return JsonEncoder.fromBase64(data.base64)
+    } else if (data.json) {
+      return data.json
+    } else {
+      return null
+    }
+  }
```

Comment:
Can we extract these as separate util method? 

Also can we align this with how attachmetns work in v1? can we reuse the same attachment class, or at least make sure there's consistency in how they work. This now redefines the attachment type in a different way. It would be nice if DIDComm v1 and v2 works as simliar as possible where it's not needed to diverge (and I think differences in attachments can be mostly abstracted away)